File: bootstrap/src/main/resources/executor/pi/test/agent-state-v2-codec.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import test from "node:test";

import {
  canonicalizeStateV2,
  decodeAndVerifyStateV2,
  hashStateV2,
} from "../src/agent-state-v2-codec.mjs";

const fixturePath = resolve(
  import.meta.dirname,
  "../../../../../../../test-fixtures/protocol/rd-agent-state-v2-canonical-fixtures.json",
);
const fixtures = JSON.parse(await readFile(fixturePath, "utf8"));

test("matches Java shared canonical and hash fixtures", () => {
  for (const fixture of fixtures.valid) {
    assert.equal(canonicalizeStateV2(fixture.input), fixture.canonical, fixture.name);
    assert.equal(hashStateV2(fixture.input), fixture.hash, fixture.name);
    assert.deepEqual(decodeAndVerifyStateV2(fixture.canonical, fixture.hash), fixture.input, fixture.name);
  }
});

test("rejects shared unsafe, malformed, and hash mismatch fixtures", () => {
  for (const fixture of fixtures.invalid) {
    assert.throws(
      () => decodeAndVerifyStateV2(fixture.json, fixture.hash),
      (error) => String(error.message).toLowerCase().includes(fixture.error),
      fixture.name,
    );
  }
});

File: rag/src/test/java/com/wish/rd/rag/project/agent/model/AgentStateV2CodecTest.java
package com.wish.rd.rag.project.agent.model;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.nio.file.Files;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class AgentStateV2CodecTest {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Test
    void matchesSharedCanonicalAndHashFixtures() throws Exception {
        JsonNode fixtures = fixtures();
        for (JsonNode fixture : fixtures.path("valid")) {
            JsonNode input = fixture.path("input");
            assertEquals(fixture.path("canonical").textValue(), AgentStateV2Codec.canonicalize(input));
            assertEquals(fixture.path("hash").textValue(), AgentStateV2Codec.hash(input));
            assertEquals(
                    input,
                    AgentStateV2Codec.decodeAndVerify(
                            fixture.path("canonical").textValue(),
                            fixture.path("hash").textValue()
                    )
            );
        }
    }

    @Test
    void rejectsSharedInvalidFixtures() throws Exception {
        for (JsonNode fixture : fixtures().path("invalid")) {
            IllegalArgumentException error = assertThrows(
                    IllegalArgumentException.class,
                    () -> AgentStateV2Codec.decodeAndVerify(
                            fixture.path("json").textValue(),
                            fixture.path("hash").textValue()
                    ),
                    fixture.path("name").textValue()
            );
            assertTrue(
                    error.getMessage().toLowerCase().contains(fixture.path("error").textValue()),
                    () -> fixture.path("name").textValue() + ": " + error.getMessage()
            );
        }
    }

    private static JsonNode fixtures() throws Exception {
        Path root = Path.of(System.getProperty("user.dir")).toAbsolutePath();
        Path fixture = root.resolve("test-fixtures/protocol/rd-agent-state-v2-canonical-fixtures.json");
        if (!Files.exists(fixture)) {
            fixture = root.getParent().resolve(
                    "test-fixtures/protocol/rd-agent-state-v2-canonical-fixtures.json"
            );
        }
        return OBJECT_MAPPER.readTree(Files.readString(fixture));
    }
}
